moved Firebase reads for markets and marketdata pages to backend

declansullivan opened the pull request:
Both markets.js and marketdata.js have been updated to have all Firebase work done in the routes section of the codebase.  

Closes #59 
Closes #60

sumeet-bansal commented on routes/markets.js:
```diff
@@ -1,8 +1,42 @@
+			var zip = childData.marketInfo.zip;
+			var level = childData.marketInfo.marketLevel;
+			
+			var markup = "<tr><td>" + "<a id='turnblue' onclick='redirectFunction(this)'>"+name+"</a>" + "</td><td>" + address + 
```
use `href` instead of `onclick` when using anchor tags and delete `redirectFunction()`

sumeet-bansal commented on routes/marketdata.js:
```diff
@@ -1,8 +1,79 @@
+                    text_muted_3 += "<br>";
+                }
+            }
+		});
```
nit: indentation

declansullivan commented on routes/config.js:
```diff
@@ -0,0 +1,10 @@
+config = {
```
https://github.com/TritonSE/live-weller/blob/master/public/js/config.js

The API key has been committed to the repo and public for months, that's why https://liveweller.herokuapp.com/ works at the moment.

sumeet-bansal replied:
Ok, just create an issue for hiding the API key and we’ll fix it later with a `.env`.

declansullivan replied:
Will do

declansullivan commented:
Marketdata templates have been added, this is good to go for review.

sumeet-bansal commented on routes/marketdata.js:
```diff
@@ -1,8 +1,58 @@
+    const marketName = req.params.marketName;
+
+    marketsRef.once('value', function(snapshot) {
+        snapshot.forEach(function(childSnapshot) {
```
This snippet looks like it’s pulling the entire database and iterating through all the market documents until it finds the correct one, which is pretty inefficient since it adds an order of magnitude to both performance and networking. We only need a single document, so we should use market ID as the route parameter instead of market name. Whenever a market is created, it should have a UUID (Universally Unique ID) generated for it, where the UUID is the market document’s key in the collection/snapshot, so we can specifically get that document instead of iterating through all the documents. This ties into having a proper schema though, so maybe we can leave this till that PR.

declansullivan replied:
Having a different PR for that would probably be good.  I figured out how to access the market info, but passing the name and address of the market that is currently used as the key through the url causes issues, especially when the address contains a pound, #, sign.

I know what to do when this comes up, and I'll get it done next.

sumeet-bansal replied:
Yeah, I was thinking about possible solutions for that issue. Maybe just replace all the problem characters with an asterisk and do a regex check? Or hash the market name and address to a safe character list (e.g. convert the string to a hexadecimal one)?

declansullivan replied:
That's what I have right now to a degree, but a little inconsistency caused some issues.  I can likely make some adjustments here and there to make it work with instant access with the current key setup, although it would probably be better just to overhaul the whole thing and get the UUID method implemented.

Hashing the market name is a good idea, I'll make some notes on it tomorrow so I know which bits to fix up.

sumeet-bansal commented on routes/markets.js:
```diff
@@ -1,8 +1,36 @@
+			let zip = childData.marketInfo.zip;
+			let level = childData.marketInfo.marketLevel;
+
+			markets.push({name: name, address: address, size: size, zip: zip, level: level});
```
You can fold the variable declarations and JSON object into one: 
```javascript
markets.push({
    name: childData.marketInfo.marketName
});
```

Also, if the key name and value variable are the same, e.g. `{ level: level }`, you don’t have to specify the value (`{ level: level }` is equivalent to `{ level }`).

sumeet-bansal commented on views/markets.ejs:
```diff
@@ -66,12 +66,23 @@
+                            <td><%= market.size %></td>
+                            <td><%= market.zip %></td>
+                            <td><%= market.level %></td>
+                            <td><button class="mapButton" onclick="window.open('https\://www.google.com/maps/dir/?api=1&destination=<%= market.address %>')"><p>Go!</p></button></td>
```
use `href` instead of `onClick`

sumeet-bansal commented on routes/marketdata.js:
```diff
@@ -1,8 +1,58 @@
+                    missed.push({key: childData.missedQuestions[key].replace('<span class=\"boldanswer\">', '').replace('</span>', '')});
+                }
+
+                res.render('marketdata', {name: marketName, level: level, address: address, size: size, first: firstName, last: lastName, email: email, status: marketStatus, questions: questions, missed: missed});
```
see the other comment about JSON objects

declansullivan commented:
Added new changes, adjusted the json declaration to just reference the market information straight from the existing json as advised, as well as changed that one onclick to an href by adding an extra <a> tag.

As for the UUID, I have ideas on how to go about it, but that is definitely suited for a different PR since it'll require restructuring how the markets are added and read from the database in a good amount of places.

sumeet-bansal commented on routes/marketdata.js:
```diff
@@ -1,8 +1,61 @@
+                    status = "Market is at Top Level!";
+                }
+
+                for (let key in childData.questions) {
```
nit: `key` can be a constant

declansullivan replied:
I'll fix this in a future PR.

sumeet-bansal commented on routes/marketdata.js:
```diff
@@ -1,8 +1,61 @@
+        snapshot.forEach(function(childSnapshot) {
+            const childData = childSnapshot.val();
+            
+            let status = "";
```
in general, it's good to declare variables as late as possible (or to reframe that, as close to its initial use as possible)

sumeet-bansal commented on routes/marketdata.js:
```diff
@@ -1,8 +1,61 @@
+
+                for (let key in childData.questions) {
+                    if (key !== "undefined") {
+                        questions.push({key: key, answer: childData.questions[key]});
```
This block (and the `childData.missedQuestions` one below) could be a little tighter by using built-in functional programming concepts like [`filter`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter) and [`map`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map):
```javascript
// so this block...
let questions = [];
for (const key in childData.questions) {
    if (key !=== "undefined") {
        questions.push({ key, answer: childData.questions[key] });
    }
}

// ...can be refactored into this
const questions = childData.questions
    .filter(key => key !== "undefined")
    .map(key => { key, answer: childData.questions[key] });
```

Not 100% sure the above's correct but it's in the ballpark.

declansullivan replied:
I'll keep that in mind and give it an edit, thanks!
